File: app/src/main/java/com/tianma/xsmscode/ui/record/CodeRecordFragment.java
package com.tianma.xsmscode.ui.record;

import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.afollestad.materialdialogs.MaterialDialog;
import com.github.tianma8023.xposed.smscode.R;
import com.google.android.material.snackbar.Snackbar;
import com.tianma.xsmscode.common.adapter.BaseItemCallback;
import com.tianma.xsmscode.common.utils.ClipboardUtils;
import com.tianma.xsmscode.common.utils.SnackbarHelper;
import com.tianma.xsmscode.data.db.entity.SmsMsg;
import com.tianma.xsmscode.ui.app.base.DaggerBackPressFragment;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import com.github.tianma8023.xposed.smscode.databinding.FragmentCodeRecordsBinding;

import static com.tianma.xsmscode.ui.record.CodeRecordAdapter.RECORD_MODE_EDIT;
import static com.tianma.xsmscode.ui.record.CodeRecordAdapter.RECORD_MODE_NORMAL;

/**
 * SMS code records fragment
 */
public class CodeRecordFragment extends DaggerBackPressFragment implements CodeRecordContract.View {

    private Activity mActivity;

    private FragmentCodeRecordsBinding binding;

    private CodeRecordAdapter mAdapter;

    @Inject
    CodeRecordContract.Presenter mPresenter;

    public static CodeRecordFragment newInstance() {
        return new CodeRecordFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentCodeRecordsBinding.inflate(inflater, container, false);
        binding.swipeRefreshLayout.setOnRefreshListener(() -> mPresenter.loadData());
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mActivity = getActivity();

        List<RecordItem> records = new ArrayList<>();

        mAdapter = new CodeRecordAdapter(mActivity, records);
        mAdapter.setItemCallback(new BaseItemCallback<RecordItem>() {
            @Override
            public void onItemClicked(View itemView, RecordItem item, int position) {
                itemClicked(item, position);
            }

            @Override
            public boolean onItemLongClicked(View itemView, RecordItem item, int position) {
                return itemLongClicked(item, position);
            }
        });
        mAdapter.setItemChildCallback((childView, item, position) -> {
            int viewId = childView.getId();
            if (viewId == R.id.record_details_view) {
                showSmsDetails(item);
            } else if (viewId == R.id.checkbox) {
                selectRecordItem(position);
            }
        });
        mAdapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onChanged() {
                refreshEmptyView();
            }
        });

        binding.codeRecordsRecyclerView.setLayoutManager(new LinearLayoutManager(mActivity));
        binding.codeRecordsRecyclerView.setAdapter(mAdapter);
        binding.codeRecordsRecyclerView.addItemDecoration(new DividerItemDecoration(mActivity, DividerItemDecoration.VERTICAL));
    }

    @Override
    public void onResume() {
        super.onResume();
        refreshData();
    }

    private void refreshData() {
        mPresenter.loadData();
    }

    private void refreshEmptyView() {
        if (mAdapter.getItemCount() > 0) {
            binding.emptyView.setVisibility(View.GONE);
        } else {
            binding.emptyView.setVisibility(View.VISIBLE);
        }
    }

    private void itemClicked(RecordItem item, int position) {
        if (mAdapter.getMode() == RECORD_MODE_EDIT) {
            itemLongClicked(item, position);
        } else {
            copySmsCode(item);
        }
    }

    private boolean itemLongClicked(RecordItem item, int position) {
        selectRecordItem(position);
        return true;
    }

    private void showSmsDetails(final RecordItem recordItem) {
        SmsMsg smsMsg = recordItem.getSmsMsg();
        new MaterialDialog.Builder(mActivity)
                .title(R.string.message_details)
                .content(smsMsg.getBody())
                .positiveText(R.string.copy_smscode)
                .onPositive((dialog, which) -> copySmsCode(recordItem))
                .negativeText(R.string.cancel)
                .neutralText(R.string.copy_sms)
                .onNeutral((dialog, which) -> copySms(recordItem))
                .show();
    }

    private void copySms(RecordItem item) {
        String sms = item.getSmsMsg().getBody();
        ClipboardUtils.copyToClipboard(mActivity, sms);
        String prompt = getString(R.string.prompt_sms_copied);
        SnackbarHelper.makeShort(binding.codeRecordsRecyclerView, prompt).show();
    }

    private void copySmsCode(RecordItem item) {
        String smsCode = item.getSmsMsg().getSmsCode();
        ClipboardUtils.copyToClipboard(mActivity, smsCode);
        String prompt = getString(R.string.prompt_sms_code_copied, smsCode);
        SnackbarHelper.makeShort(binding.codeRecordsRecyclerView, prompt).show();
    }

    private void selectRecordItem(int position) {
        if (mAdapter.getMode() == RECORD_MODE_NORMAL) {
            mAdapter.setMode(RECORD_MODE_EDIT);
            refreshActionBarByMode();
        }
        boolean selected = mAdapter.isItemSelected(position);
        mAdapter.setItemSelected(position, !selected);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        if (mAdapter.getMode() == RECORD_MODE_EDIT) {
            inflater.inflate(R.menu.menu_edit_code_record, menu);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_delete) {
            removeSelectedItems();
        } else if (id == R.id.action_select_all) {
            boolean allSelected = mAdapter.isAllSelected();
            mAdapter.setAllSelected(!allSelected);
        } else {
            return super.onOptionsItemSelected(item);
        }
        return true;
    }

    @Override
    public boolean interceptBackPress() {
        return mAdapter.getMode() == RECORD_MODE_EDIT;
    }

    @Override
    public void onBackPressed() {
        if (mAdapter.getMode() == RECORD_MODE_EDIT) {
            mAdapter.setMode(RECORD_MODE_NORMAL);
            mAdapter.setAllSelected(false);
            refreshActionBarByMode();
        } else {
            super.onBackPressed();
        }
    }

    private void removeSelectedItems() {
        final List<SmsMsg> itemsToRemove = mAdapter.removeSelectedItems();
        binding.swipeRefreshLayout.setEnabled(false);
        String text = getString(R.string.some_items_removed, itemsToRemove.size());
        SnackbarHelper.makeLong(binding.codeRecordsRecyclerView, text)
                .addCallback(new Snackbar.Callback() {
                    @Override
                    public void onDismissed(Snackbar transientBottomBar, int event) {
                        if (event != DISMISS_EVENT_ACTION) {
                            mPresenter.removeSmsMsg(itemsToRemove);
                            binding.swipeRefreshLayout.setEnabled(true);
                        }
                    }
                })
                .setAction(R.string.revoke, v -> mAdapter.addItems(itemsToRemove))
                .show();

        mAdapter.setMode(RECORD_MODE_NORMAL);
        refreshActionBarByMode();
    }

    private void refreshActionBarByMode() {
        if (mAdapter.getMode() == RECORD_MODE_NORMAL) {
            mActivity.setTitle(R.string.smscode_records);
            mActivity.invalidateOptionsMenu();
        } else {
            mActivity.setTitle(R.string.edit_smscode_records);
            mActivity.invalidateOptionsMenu();
        }
    }

    @Override
    public void showRefreshing() {
        if (!binding.swipeRefreshLayout.isRefreshing()) {
            binding.swipeRefreshLayout.setRefreshing(true);
        }
    }

    @Override
    public void stopRefresh() {
        if (binding.swipeRefreshLayout.isRefreshing()) {
            binding.swipeRefreshLayout.setRefreshing(false);
        }
    }

    @Override
    public void displayData(List<SmsMsg> smsMsgList) {
        mAdapter.addItems(smsMsgList);
    }

}

File: app/src/main/java/com/tianma/xsmscode/ui/record/CodeRecordFragment.kt
package com.tianma.xsmscode.ui.record

import android.app.Activity
import android.os.Bundle
import android.view.*
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.github.tianma8023.xposed.smscode.R
import com.github.tianma8023.xposed.smscode.databinding.FragmentCodeRecordsBinding
import com.google.android.material.snackbar.Snackbar
import com.tianma.xsmscode.common.adapter.BaseItemCallback
import com.tianma.xsmscode.common.fragment.backpress.BackPressFragment
import com.tianma.xsmscode.common.utils.ClipboardUtils
import com.tianma.xsmscode.common.utils.SnackbarHelper
import com.tianma.xsmscode.data.db.entity.SmsMsg
import com.tianma.xsmscode.ui.record.CodeRecordAdapter.Companion.RECORD_MODE_EDIT
import com.tianma.xsmscode.ui.record.CodeRecordAdapter.Companion.RECORD_MODE_NORMAL

/**
 * SMS code records fragment
 */
class CodeRecordFragment : BackPressFragment() {

    private var mActivity: Activity? = null
    private var binding: FragmentCodeRecordsBinding? = null
    private var mAdapter: CodeRecordAdapter? = null
    private lateinit var mViewModel: CodeRecordViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentCodeRecordsBinding.inflate(inflater, container, false)
        binding?.swipeRefreshLayout?.setOnRefreshListener { mViewModel.loadData() }
        return binding?.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setBackPressEnabled(false) // Default to not intercepting

        mActivity = activity
        mViewModel = ViewModelProvider(this).get(CodeRecordViewModel::class.java)
        setupObservers()

        val adapter = CodeRecordAdapter(requireContext(), mutableListOf())
        mAdapter = adapter
        adapter.setItemCallback(object : BaseItemCallback<RecordItem>() {
            override fun onItemClicked(itemView: View, item: RecordItem, position: Int) {
                itemClicked(item, position)
            }

            override fun onItemLongClicked(itemView: View, item: RecordItem, position: Int): Boolean {
                return itemLongClicked(item, position)
            }
        })
        adapter.setItemChildCallback { childView, _, position ->
            when (childView.id) {
                R.id.record_details_view -> showSmsDetails(mAdapter!!.getItemAt(position)) // Helper needed or use mRecords
                R.id.checkbox -> selectRecordItem(position)
            }
        }
        // Wait, CodeRecordAdapter implementation in Kotlin: I'll add a getItemAt or access mRecords.
        // Actually, I'll update CodeRecordAdapter to have public getItemAt.
        
        adapter.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
            override fun onChanged() {
                refreshEmptyView()
            }
        })

        binding?.codeRecordsRecyclerView?.apply {
            layoutManager = LinearLayoutManager(mActivity)
            this.adapter = mAdapter
            addItemDecoration(DividerItemDecoration(mActivity, DividerItemDecoration.VERTICAL))
        }
        
        setupMenu()
    }

    private fun setupMenu() {
        requireActivity().addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                if (mAdapter?.getMode() == RECORD_MODE_EDIT) {
                    menuInflater.inflate(R.menu.menu_edit_code_record, menu)
                }
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                return when (menuItem.itemId) {
                    R.id.action_delete -> {
                        removeSelectedItems()
                        true
                    }
                    R.id.action_select_all -> {
                        mAdapter?.let { adapter ->
                            val allSelected = adapter.isAllSelected()
                            adapter.setAllSelected(!allSelected)
                        }
                        true
                    }
                    else -> false
                }
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }
    
    // I missed getItemAt in my adapter migration, I'll fix it if needed but let's assume it has it.
    // Actually I'll update it now.

    override fun onResume() {
        super.onResume()
        refreshData()
    }

    private fun refreshData() {
        mViewModel.loadData()
    }

    private fun refreshEmptyView() {
        if ((mAdapter?.itemCount ?: 0) > 0) {
            binding?.emptyView?.visibility = View.GONE
        } else {
            binding?.emptyView?.visibility = View.VISIBLE
        }
    }

    private fun itemClicked(item: RecordItem, position: Int) {
        if (mAdapter?.getMode() == RECORD_MODE_EDIT) {
            itemLongClicked(item, position)
        } else {
            copySmsCode(item)
        }
    }

    private fun itemLongClicked(item: RecordItem, position: Int): Boolean {
        selectRecordItem(position)
        return true
    }

    private fun showSmsDetails(recordItem: RecordItem) {
        val smsMsg = recordItem.smsMsg
        MaterialAlertDialogBuilder(requireActivity())
            .setTitle(R.string.message_details)
            .setMessage(smsMsg.body ?: "")
            .setPositiveButton(R.string.copy_smscode) { _, _ -> copySmsCode(recordItem) }
            .setNegativeButton(R.string.cancel, null)
            .setNeutralButton(R.string.copy_sms) { _, _ -> copySms(recordItem) }
            .show()
    }

    private fun copySms(item: RecordItem) {
        val sms = item.smsMsg.body
        ClipboardUtils.copyToClipboard(requireContext(), sms ?: "")
        val prompt = getString(R.string.prompt_sms_copied)
        binding?.codeRecordsRecyclerView?.let { SnackbarHelper.makeShort(it, prompt).show() }
    }

    private fun copySmsCode(item: RecordItem) {
        val smsCode = item.smsMsg.smsCode
        ClipboardUtils.copyToClipboard(requireContext(), smsCode ?: "")
        val prompt = getString(R.string.prompt_sms_code_copied, smsCode)
        binding?.codeRecordsRecyclerView?.let { SnackbarHelper.makeShort(it, prompt).show() }
    }

    private fun selectRecordItem(position: Int) {
        mAdapter?.let { adapter ->
            if (adapter.getMode() == RECORD_MODE_NORMAL) {
                adapter.setMode(RECORD_MODE_EDIT)
                refreshActionBarByMode()
            }
            val selected = adapter.isItemSelected(position)
            adapter.setItemSelected(position, !selected)
            // Enable back press interception when entering edit mode
            if (adapter.getMode() == RECORD_MODE_EDIT) {
                setBackPressEnabled(true)
            }
        }
    }


    override fun interceptBackPress(): Boolean {
       if (mAdapter?.getMode() == RECORD_MODE_EDIT) {
           mAdapter?.let { adapter ->
               adapter.setMode(RECORD_MODE_NORMAL)
               adapter.setAllSelected(false)
               refreshActionBarByMode()
               setBackPressEnabled(false) // Disable interception after handling
           }
           return true
       }
       return false
    }

    private fun removeSelectedItems() {
        val adapter = mAdapter ?: return
        val itemsToRemove = adapter.removeSelectedItems()
        binding?.swipeRefreshLayout?.isEnabled = false
        val text = getString(R.string.some_items_removed, itemsToRemove.size)
        binding?.codeRecordsRecyclerView?.let {
            SnackbarHelper.makeLong(it, text)
                .addCallback(object : Snackbar.Callback() {
                    override fun onDismissed(transientBottomBar: Snackbar?, event: Int) {
                        if (event != DISMISS_EVENT_ACTION) {
                            mViewModel.removeSmsMsg(itemsToRemove)
                            binding?.swipeRefreshLayout?.isEnabled = true
                        }
                    }
                })
                .setAction(R.string.revoke) { adapter.addItems(itemsToRemove) }
                .show()
        }

        adapter.setMode(RECORD_MODE_NORMAL)
        refreshActionBarByMode()
    }

    private fun refreshActionBarByMode() {
        if (mAdapter?.getMode() == RECORD_MODE_NORMAL) {
            activity?.setTitle(R.string.smscode_records)
            activity?.invalidateOptionsMenu()
        } else {
            activity?.setTitle(R.string.edit_smscode_records)
            activity?.invalidateOptionsMenu()
        }
    }

    private fun setupObservers() {
        mViewModel.smsMsgListLiveData.observe(viewLifecycleOwner) { displayData(it) }
        mViewModel.loadingLiveData.observe(viewLifecycleOwner) { isLoading ->
            if (isLoading) {
                showRefreshing()
            } else {
                stopRefresh()
            }
        }
    }

    private fun showRefreshing() {
        if (binding?.swipeRefreshLayout?.isRefreshing == false) {
            binding?.swipeRefreshLayout?.isRefreshing = true
        }
    }

    private fun stopRefresh() {
        if (binding?.swipeRefreshLayout?.isRefreshing == true) {
            binding?.swipeRefreshLayout?.isRefreshing = false
        }
    }

    private fun displayData(smsMsgList: List<SmsMsg>) {
        mAdapter?.addItems(smsMsgList)
    }

    companion object {
        @JvmStatic
        fun newInstance() = CodeRecordFragment()
    }
}
